File: src/20. Valid Parentheses/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c not in mapping:
                stack.append(c)
            elif not stack or mapping[c] != stack.pop():
                return False
        return not stack


if __name__ == "__main__":
    s = "()[]{}"
    print(Solution().isValid(s))

File: src/20. Valid Parentheses/20.go
func isValid(s string) bool {
    stack := make([]byte, 0, len(s))

    for _, c := range []byte(s) {
        if c == '(' || c == '[' || c == '{' {
            stack = append(stack, c)
            continue
        }

        if len(stack) == 0 {
            return false
        }

        if c == ')' && stack[len(stack)-1] != '(' ||
           c == ']' && stack[len(stack)-1] != '[' ||
           c == '}' && stack[len(stack)-1] != '{' {
                return false
        }
        stack = stack[:len(stack)-1]
    }

    return len(stack) == 0    
}